RFC-0017-PyTorch-Operator-Versioning.md

```diff
@@ -0,0 +1,223 @@
+
+This document proposes a new BC and FC policy based on operator versioning. 
+
+
+## History
+
+
+### Backwards Compatibility
+
+Backwards compatibility (BC), the ability for PyTorch to continue running programs from older versions, is important so programs don’t need to be rewritten.
+
+PyTorch makes the following BC promises today:
+
+
+
+* **OSS** — “stable” features will be deprecated for one release before a BC-breaking change is made. [PyTorch OSS BC-breaking policy](https://pytorch.org/docs/master/) 
+* **FB Internal** — we will not break a serialized torchscript program running in production at Facebook
+
+BC-breaking operator changes were previously governed by the [Backward-compatibility Breaking Change Review Process](https://fb.quip.com/gydOArylrcKd), but this only covered torchscript and eager.
+
+
+### Forwards Compatibility
+
+Forwards compatibility (BC), the ability for older versions of PyTorch to run programs from newer versions, is important so users don’t need to update PyTorch.
+
+PyTorch makes the following FC promises today:
+
+
+
+* **OSS** — no promise
+* **FB Internal** — PyTorch commits can run existing PyTorch eager, package/deploy, and serialized torchscript programs for at least two weeks
+    * The addition of a new kwarg-only argument at the end of an op’s parameter list (but before out=, if present) with a default value is FC-compatible for serialized [torchscript](https://fb.workplace.com/groups/pytorch.dev/permalink/909079013003913/) and [edge](https://fb.workplace.com/groups/pytorch.dev/permalink/912379562673858/).
+
+
+## Goals
+
+
+
+* To support backward and (some) forward compatibility for an arbitrary BC or FC break update (schema, functional, etc) on an operator by [versioning](https://docs.google.com/document/d/1nyXmss2O003ZgKrhDmd-kyLNjjMqEXww_2skOXqkks4/edit). 
+* It’s a systematic flow to prevent BC/FC breakage on both deploy and runtime stages.
+* To provide testing that accurately detects dangerous BC and FC-breaking changes.
+* To support both server and edge use cases, including TorchScript, package/deploy and Edge.
+
+
+## Non-goals
+
+
+
+* It does not mean that models with old operator schema can **always** run successfully on new runtime and vice versa. 
+* It’s not for the “automatic” BC/FC support that can be done without any developer’s manual work (for example, the number of arguments mentioned in the Context). To apply versioning on the updated operator, the author needs to manually add a line in the version table and provide the resolution function for BC. This proposal is for BC/FC breakages that the automatic supports don’t apply. 
+* It does not include the BC/FC for package/deploy itself. The Python-only operators are transparent to TS and Edge clients, with the TS compilation. 
+
+
+# Proposal
+
```

Comment:
Good catch. I thought I have added @mruberry 's nice glossary. Now it's added. Also add a line of "We propose the operator versioning that works across eager, TorchScript, torch.package and mobile. It uses a version number + corresponding upgraders in torchscript to avoid breakage due to BC/FC breaking operator updates.". Please let me know they look good.